### PR TITLE
Implement AssignAllUnits() and MakeCurrent() state/modelgeneration

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

This patch implements the following methods on the `state.Generation` receiver:
- `AssignAllUnits`, for use when the `set-generation` command is used.
- `MakeCurrent`, used for verifying the 'next' generation can be completed and marking it so.  For use when auto completion of a generation is possible with `set-generation.

## QA steps

No activated logic yet. Accompanying unit tests verify correctness.

## Documentation changes

Ultimately, yes.

## Bug reference

N/A

### CODE REVIEW

File: state/modelgeneration.go [hunk 1]
```diff
@@ -172,6 +172,55 @@ func assignGenerationUnitTxnOps(id, appName, unitName string) []txn.Op {
 	}
 }
 
+// Complete marks the generation as completed.
+func (g *Generation) Complete() error {
```
Comment (hmlanigan) [inline state/modelgeneration.go]: Is there a better name for this?

File: state/modelgeneration.go [hunk 2]
```diff
@@ -172,6 +172,55 @@ func assignGenerationUnitTxnOps(id, appName, unitName string) []txn.Op {
...
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ops := []txn.Op{
+			{
+				C:      generationsC,
+				Id:     g.doc.Id,
+				Assert: bson.D{{"$and", assertAndCriteria}},
+				Update: bson.D{
+					{"$set", bson.D{{"completed", time.Unix()}}},
```
Comment (hmlanigan) [inline state/modelgeneration.go]: @manadart, Is time what you had in mind for completed, once it's completed?  Or something else?

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	for i, entity := range args.Entities {
+		tag, err := names.ParseTag(entity.Tag)
+		if err != nil {
+			results.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		switch tag.Kind() {
+		case names.ApplicationTagKind:
+			unitNames, err := m.state.AppUnitNames(tag.String())
```
Comment (hmlanigan) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: Move this stanza to generation.AssignAllUnits()?  Then appUnitNames() can stay private.

File: state/modelgeneration.go (hunk 2 above)
Comment (manadart) [reply state/modelgeneration.go]: Yes, exactly this.

File: state/modelgeneration.go (hunk 1 above)
Comment (manadart) [reply state/modelgeneration.go]: `MakeCurrent`, `Activate`, `Apply`? I think the first is good verbiage for what we are doing.

File: apiserver/facades/client/modelgeneration/modelgeneration.go (hunk 3 above)
Comment (manadart) [reply apiserver/facades/client/modelgeneration/modelgeneration.go]: Yes, I would push it in there.

File: state/modelgeneration.go [hunk 4]
```diff
@@ -171,6 +171,55 @@ func assignGenerationUnitTxnOps(id, appName, unitName string) []txn.Op {
...
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt > 0 {
+			if err := g.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if !g.Active() {
+			return nil, errors.New("next generation is not currently active")
+		}
+		ok, err := g.CanAutoComplete()
```
Comment (manadart) [inline state/modelgeneration.go]: If we choose another name for `Complete` we should change this name too, along with the error prose below.

File: state/modelgeneration.go (hunk 1 above)
Comment (hmlanigan) [reply state/modelgeneration.go]: MakeCurrent works for me.  ty.

File: state/modelgeneration.go (hunk 4 above)
Comment (hmlanigan) [reply state/modelgeneration.go]: done

File: apiserver/facades/client/modelgeneration/modelgeneration.go (hunk 3 above)
Comment (hmlanigan) [reply apiserver/facades/client/modelgeneration/modelgeneration.go]: done

File: state/modelgeneration.go [hunk 5]
```diff
@@ -171,24 +218,72 @@ func assignGenerationUnitTxnOps(id, appName, unitName string) []txn.Op {
...
+		}
+		if !g.Active() {
+			return nil, errors.New("generation is not currently active")
+		}
+		ok, err := g.CanMakeCurrent()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if !ok {
+			return nil, errors.New("generation can not auto complete")
```
Comment (manadart) [inline state/modelgeneration.go]: Maybe just "generation can not be completed" or "generation can not be applied"?

File: state/modelgeneration.go [hunk 6]
```diff
@@ -87,12 +88,11 @@ func (g *Generation) AssignApplication(appName string) error {
 		if _, ok := g.doc.AssignedUnits[appName]; ok {
 			return nil, jujutxn.ErrNoOperations
 		}
+		// Any 'next' generation that is Active, cannot also be Completed,
+		// see MakeCurrent() and NextGeneration().
 		if !g.Active() {
 			return nil, errors.New("generation is not currently active")
 		}
-		if g.doc.Completed > 0 {
```
Comment (manadart) [inline state/modelgeneration.go]: Nice; this makes completed -> inactive work for us.

File: state/modelgeneration.go (hunk 5 above)
Comment (hmlanigan) [reply state/modelgeneration.go]: i went with the former

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
fail in: UniterSuite.TestUniterRelations

$$merge$$

**Pull request merged**